File: solutions/code-carl/ch05_stack/T7_LC0347_TopKFrequent.java
/**
 * Practice template for:
 * - LeetCode 347. Top K Frequent Elements
 *   https://leetcode.cn/problems/top-k-frequent-elements/
 */
import java.util.HashMap;
import java.util.PriorityQueue;

public class T7_LC0347_TopKFrequent {
    public int[] topKFrequent(int[] nums, int k) {
        HashMap<Integer, Integer> map = new HashMap<>();
        for (int num : nums) {
            map.put(num, map.getOrDefault(num, 0) + 1);
        }

        PriorityQueue<Integer> queue = new PriorityQueue<>((key1, key2) -> map.get(key1) - map.get(key2));
        for (Integer key : map.keySet()) {
            queue.offer(key);
            if (queue.size() > k) {
                queue.remove();
            }
        }
        return queue.stream().mapToInt(Integer::intValue).toArray();
    }
}

File: solutions/python/code-carl/ch05_stack/t7_lc0347_top_k_frequent.py
"""LeetCode 347. Top K Frequent Elements
https://leetcode.cn/problems/top-k-frequent-elements/
"""
import heapq
from collections import Counter
from typing import List


class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        cnt = Counter(nums)
        heap = []

        for num, freq in cnt.items():
            if len(heap) < k:
                heapq.heappush(heap, (freq, num))
            else:
                if freq > heap[0][0]:
                    heapq.heapreplace(heap, (freq, num))

        return [num for _, num in heap]


if __name__ == "__main__":
    solution = Solution()
    result1 = solution.topKFrequent([1, 1, 1, 2, 2, 3], 2)
    assert set(result1) == {1, 2}

    result2 = solution.topKFrequent([1], 1)
    assert result2 == [1]

    print("All tests passed.")
